python/tvm/relay/backend/graph_runtime_codegen.py

```diff
@@ -348,13 +348,39 @@ def _get_json(self):
             attrs["device_index"] = ["list_int", device_types]
         attrs["dltype"] = ["list_str", dltypes]
 
-        json_dict = {
+        # Metadata definitions
+        graph_inputs = {}
+        graph_outputs = {}
+        for node_id in arg_nodes:
+            node_name = nodes[node_id]['name']
+            if node_name not in self.params:
+                graph_inputs[node_name] = {}
+                graph_inputs[node_name]['id'] = node_id
+                graph_inputs[node_name]['dtype'] = dltypes[node_id]
+                graph_inputs[node_name]['shape'] = shapes[node_id]
+        for node_id in heads:
+            node_name = nodes[node_id[0]]['name']
+            graph_outputs[node_name] = {}
+            graph_outputs[node_name]['id'] = node_id[0]
+            graph_outputs[node_name]['dtype'] = dltypes[node_id[0]]
+            graph_outputs[node_name]['shape'] = shapes[node_id[0]]
+
+        metadata = {}
```

Comment:
FWIW if you're using collections, this can just be:

```
def nested_defaultdict():
  return defaultdict(nested_defaultdict)

metadata = nested_defaultdict()
metadata['signatures']['default'] = {'inputs': graph_inputs, 'outputs': graph_outputs}
```